Operation Workflow Specification & Implementation (#2071)
thin-edge/thin-edge.io: docs/src/references/operation-workflow.md

---
title: Operation Workflows
tags: [Reference, MQTT, Operations]
sidebar_position: 6
---

# Operation Workflows

Thin-edge provides the tools to define, extend and combine *operation workflows*
that rule the sequence of steps applied when a maintenance *operation* is triggered by an operator or some software component,
whether it is a *command* to restart the device, to update a configuration file or to install a new software.

An operation workflow specifies the timeline of the actions to perform
as well as any prerequisite checks, outcome validations and possible rollbacks.
However, a workflow doesn't define how to perform these actions.
These are delegated to software components participating in the operation progress.
More precisely, an operation workflow defines:
- the *observable states* of an ongoing operation instance
  from initialization up to a final success or failure
- the *participants* and their interactions, passing the baton to the software component
  which responsibility is to advance the operation in a given state
  and to notify the other participants what is the new resulting state

> - the *possible state sequences* so the system can detect any stale or miss-behaving operation request.

Reviewer: ```suggestion
- the *possible state sequences* so the system can detect any stale or misbehaving operation request.
```

Author: Fixed

One can subscribe to the following topic to get all the capabilities of a thin-edge device and its child-devices and services.

```sh te2mqtt
tedge mqtt sub 'te/+/+/+/+/cmd/+' 
```

The retained messages published on these topics are operation specific and defined by the operation APIs.
They provide operation specific parameters such as the list of software package types that can be installed,
or the list of file types that configured.

As an example, the `tedge-agent` which implements the `restart` and `software_update` capabilities for the main device,
will emit two retained messages.

A first message to tell that the main device can be restarted:

```sh te2mqtt
tedge mqtt pub -r 'te/device/main///cmd/restart' '{}' 
```

A second one to tell that debian packages can be installed on the main device: 

```sh te2mqtt
tedge mqtt pub -r 'te/device/main///cmd/software_update' '{ "type": ["apt"] }' 
```

### Commands

The topics matching __te__/*a*/*b*/*c*/*d*/__cmd__/*operation*/*command-id* are used to trigger and manage commands,
i.e. operation requests on a specific target for a specific *operation*.

> Each request is given a unique topic where the current state of the command workflow is published as a retained message.

Author: Added

    associated by another software component responsible for these rollbacks.

Furthermore, specific versions of the same workflow can be defined on different targets.
Indeed, all the status updates for a command on a given thin-edge entity or component
are published on an MQTT topic which prefix is the entity identifier.
- The same executable can be used to handle operations on different targets.
  For instance, the `tedge-configuration-plugin` can run on the main device `te/device/main//`
  as well as on a child-device identified by `te/device/child-xyz//`.
- A specific executable can be substituted on a specific target.
  If for some reasons, `tedge-configuration-plugin` cannot be installed on a child-device,
  then a specific implementation of the `configuration-update` MQTT API can be used to serve configuration updates
  on, say, `te/micro-controller/xyz//`.
- A workflow can be extended differently for each target.
  As an example, an agent developer can define an extra rollback state on the main device but not on the child devices.

### Operation API

As several software components have to collaborate when executing a command, each operation must define a specific API.

> This API should be based on the principles of MQTT-driven workflow and defines: 

Author: Agree. I have to move the whole content under the agent reference documentation and to break this content in two parts : context (already partly here) and workflow (the contract described below).

> This API should be based on the principles of MQTT-driven workflow and defines: 

Author: The docs have been re-organized

- create new workflows, combining other workflows and steps

- A *workflow* defines the possible sequences of actions for an operation request
  from its initialization up to its success or failure.

## Operations, Capabilities, and Commands

From a user perspective an *operation* is a predefined sequence of actions
that an operator can trigger on a device to reach some desirable state.
It can be to restart the device or to install some new software.
From an implementation perspective, an operation is an API identified by a well-known name such as `restart` or `software-update`.
This API rules the coordination among the software components that need to interact to advance the operation.

Not all entities and components of a thin-edge device support all the operations,
and, even if they do, the implementations might be specific.
Installing a software package on top of service makes no sense.
Restarting the device is not the same as restarting one of its services.
Each entity or component has to declare its *capabilities* i.e. the operations made available on this target.

Strictly speaking, capabilities are not implemented nor declared by the devices and the services themselves.

> There are implemented by thin-edge services and plugins.

Reviewer 2: ```suggestion
These are implemented by thin-edge services and plugins.
```

Author: fixed

## MQTT Topics

Operations, capabilities and commands are declared, triggered and managed using MQTT topics,
all built along the same schema, matching the topic filter `te/+/+/+/+/cmd/+/+`,
with a target prefix `te/+/+/+/+` and a command specific suffix `/cmd/+/+`:

| root   | target           | command keyword | operation name | command instance id |
|--------|------------------|-----------------|----------------|---------------------|
| __te__ | /*a*/*b*/*c*/*d* | /__cmd__        | /*operation*   | /*command-id*       |

The prefix __te__/*a*/*b*/*c*/*d* uniquely identifies the entity or component that is the target of commands.
It can be:
- the main device: `te/device/main//`
- a child device: `te/device/child-xyz//`
- a service: `te/device/main/service/tedge-agent`
- or any application specific entity identifier such as `te/raspberry-pi/123/process/collectd`.

The longer prefix __te__/*a*/*b*/*c*/*d*/__cmd__ groups all the capabilities and commands
related to the entity identified by __te__/*a*/*b*/*c*/*d*.

### Capabilities

A capability, the ability for an entity __te__/*a*/*b*/*c*/*d* to handle a given *operation*, is published as a retained message

> on the topic __te__/*a*/*b*/*c*/*d*/__cmd__/*operation* which suffix is the well-known name of the operation.

Author: fixed

> One can subscribe to the following topic to get all the capabilities of a thin-edge device.

Author: updated

```sh te2mqtt
tedge mqtt sub 'te/+/+/+/+/cmd/+' 
```

The retained messages published on these topics are operation specific and defined by the operation APIs.

> They provide operation specific parameters such as the types of software packages that can be installed.

Author: Updated with a slightly different wording.

As an example, the `tedge-agent` which implements the `restart` and `software-update` capabilities for the main device,
will emit two retained messages.

A first message to tell that the main device can be restarted:

```sh te2mqtt
tedge mqtt pub -r 'te/device/main///cmd/restart' '{}' 
```

A second one to tell that debian packages can be installed on the main device: 

```sh te2mqtt
tedge mqtt pub -r 'te/device/main///cmd/software-update' '{ "type": ["apt"] }' 
```

### Commands

The topics matching __te__/*a*/*b*/*c*/*d*/__cmd__/*operation*/*command-id* are used to trigger and manage commands,
i.e. operation requests on a specific target for a specific *operation*.

Depending on the operation type, there are two different approaches to identify a command, i.e. an operation request.
- Operations that require no coordination among components use a simple scheme,
  where all the requests are published on the same topic, by convention, the `req` topic.
- Operations driven by a workflow, use a topic per command, i.e. per operation request.
  Each requests is given a unique topic where the current state of the command workflow is published as a retained message.

>   This unique id assigned by the requester, who is also responsible for creating the topic and finally removing it.

Author: added

An example, for the simple case, is the `health-check` operation.
Thin-edge services published status messages on start and stop.
The point of the `health-check` operation is to ask a service to publish a fresh status message.
No coordination is required and the outcome is observed elsewhere.

> All the requests are then simply published on the `.../cmd/health-check/req` sub-topic of the target.

Author: Added

> Here the `tedge-agent` running on the main device:

Author: Added

```sh te2mqtt
tedge mqtt pub 'te/device/main/service/tedge-agent/cmd/heath-check/req' '{}' 
```

Software update is an example of an operation that requires coordination.
On reception of a software update request from the cloud operator,
the `tedge-mapper` creates a fresh new topic for this command,
say `te/device/main///cmd/software-update/req-123` for the 123<sup>rd</sup> request.
On this topic, a first retained messages is published to describe the operator expectations for the software updates.

```sh te2mqtt
tedge mqtt pub -r 'te/device/main///cmd/software-update/req-123' '{

>     "modules": [

Author: Added

        {
            "type": "apt",
            "name": "collectd",
            "version": "5.7",
            "action": "install"
        }
    ]
}' 
```

> Then, the `tedge-agent` and possibly other software components take in charge the command, making it advance to some final state,

Reviewer 2: ```suggestion
Then, the `tedge-agent` and possibly other software components take charge of the command, making it advance to some final state,
```

Author: fixed

- The messages published over this topic represent the current state of the command.
  - Such a message states where the command progression is and gives all the required information to proceed.
  - e.g. `{ "status": "Requested", "operation": "ConfigurationUpdate", "target": "mosquitto", "url": "https://..."}`
- The state messages are published as retained.
  - They capture the latest state of the operation request.
  - Till some change occurs, this latest state is dispatched to any participant on reconnect.
- Several participants act in concert to move forward the command execution.
  - The participants observe the progress of all the operations they are interested in.
  - They watch for the specific states they are responsible in moving forward.
  - When a step is performed, successfully or not, the new state is published accordingly by the performer.

### Example

Here is an example where three software components participate in a `configuration-update` command.
- The `tedge-mapper` creates the initial state of the command
  providing the required information to install a new version for a configuration file;
  and then waits for the final outcome (in black).

> - The `tedge-config-plugin` handles the main steps (in red): downloading the file and installing it where expected.

Author: That's correct. This is defined [here](https://github.com/thin-edge/thin-edge.io/pull/2069)

> - The `tedge-config-plugin` handles the main steps (in red): downloading the file and installing it where expected.

Author: With hindsight, notably after https://github.com/thin-edge/thin-edge.io/pull/2440 and offline discussions, we plan to revise this point of view, *making the mapper no more responsible of the download steps*.

Indeed:
- Systematically downloading the resources to a local cache before triggering the operation might be a waste of time as the operation might be cancelled for independent reasons (e.g. say a custom operation rule prevents any update when there is a specific alarm). This is also a source of complexity ([download from cloud, upload to the local file transfer service, download from this service](https://github.com/thin-edge/thin-edge.io/pull/2440)).
- Bypassing the local file-transfer cache simplifies the workflow but moves on the child devices the responsibility to handle long and possibly interrupted downloads from the cloud. This also prevents caching resources targeting several child devices. 

So, a more flexible approach could be to use a configurable workflow to trigger or not intermediate downloads. 
- The role of the mapper is only to trigger and monitor the operations. It provides the required URLs to the cloud resources, but download nothing. The c8y mapper rewrites the c8y URLS into local URLs going through the c8y auth proxy, but doesn't trigger the downloads.
- The user-defined workflow for the operation may or may not trigger a download of the cloud resources to the local file transfer service. If doing so, the workflow has to *update* the initial command substituting the former cloud URLs to local URLs for the same resources.
- When its turn arrives, the operation plugin uses the URLs provided in the command payload. These can be local or cloud URLs.

As download / upload are key step for device management, we can consider a specific workflow action for that. Technically, this should even be simpler, as `tedge-agent` runs both the file transfer service and the operation workflow supervision service (the cloud resource can be downloaded in place with no need for an extra upload).

- An extra component, provided by the user, handles domain-specific checks (in blue)
  to timely schedule the command as well as to ensure the configuration file is not corrupted and properly installed.

```mermaid
stateDiagram-v2
    [*] --> Init
    Init --> Download
    Download --> Downloaded
    Downloaded --> Install
    Install --> Installed
    Installed --> Successful
    Init --> Failed
    Download --> Failed
    Downloaded --> Failed
    Install --> Failed

>     Installed --> Failed

Author: This is indeed something that can be done.

    Successful --> [*]
    Failed --> [*]
    
    classDef specific color:blue;
    class Init, Downloaded, Installed specific
    
    classDef plugin color:red;
    class Download, Install plugin
       
    classDef mapper color:black;
    class Successful, Failed mapper
```

### Benefits

Observe on the example that:

- At any state, *one and only one* participant is responsible to move the operation forward.
- Publishing a state to the MQTT command topic, can be seen as passing the baton from one participant to another.

>   The mapper creates the `Init` state and let then the other components work.

Reviewer 2: ```suggestion
  The mapper creates the `Init` state and lets the other components work.
```

Author: fixed

  The plugin tell the download has been successful by publishing the `Downloaded` state,
  but do nothing till the domain-specific component has checked the file and move the command state to `Install`.
- Each software component has to know only *some* states of the whole workflow:
  - the states they are responsible for

>   - the states they create to pass the baton.

Reviewer 2: How about using `control` instead of `baton`?

Author: changed

- The compatibility of two participants, one publishing the state owned by the other, is only defined by the message payload:
  - all the property fields expected to make progress at the new state must be provided.

The benefits are that:
- One of the participant can be substituted by another implementation as long as the substitute implementation
  is ready to process at least all the state processed by the former implementation.
  - This is the key principle used by thin-edge to provide extensible operation support.
  - The `tedge-configuration-plugin` defines the `Init`, `Downloaded`, `Installed` states
    with no specific behavior beyond proceeding to the next step;
    so, a domain specific component can be substituted to add extra checks and actions before moving forward. 

> - Extra states and participants can be added as long as each state is owned by one of the participant.

Reviewer: ```suggestion
- Extra states and participants can be added as long as each state is owned by one participant.
```

An alternative correction would be:

```
- Extra states and participants can be added as long as each state is owned by one of the participants.
```

Author: fixed

  - For instance, an agent developer can introduce a `Rollback` state in the `configuration-update` workflow,
    associated by another software component responsible for these rollbacks.

> Furthermore, specific versions of the same workflow can be defined on different targets.

Author: The same `tedge-configuration-plugin` can be installed on the main device and a child devices (same implementation different targets). The built-in workflow might have then to be overridden in a different manner on the main vs the child device (say because the operating system is not the same).

> Furthermore, specific versions of the same workflow can be defined on different targets.

Author: Not for 1.0

Indeed, all the status updates for a command on a given thin-edge entity or component
are published on an MQTT topic which prefix is the entity identifier.
- The same executable can be used to handle operations on different targets.

>   For instance, the `tedge-configuration-plugin` can run on the main device `te/device/main//`

Author: Indeed, this can also work like this.

  as well as on a child-device identified by `te/device/child-xyz//`.
- A specific executable can be substituted on a specific target.
  If for some reasons, `tedge-configuration-plugin` cannot be installed on a child-device,
  then a specific implementation of the `configuration-update` MQTT API can be used to serve configuration updates

>   on, say, `te/micro-controller/xyz//`.

Author: > It could have been the same `te/device/child-xyz//` target, right?

Sure. I just want to give a hint of why this child device is specific: this is a micro-controller where `tedge-configuration-plugin` cannot be installed.

> - A workflow can be differently extended for each target.

Author: fixed

However, there are some rules and best practices.

- All the state messages must be published as retained with QoS 1.
- A workflow implementation should not reject a state message payload with unknown fields.
  - It's also important to keep these unknown fields in the following states.
  - This is important as we want to *extend* the workflow of an operation.
    A software component added by the user might need these *extra* fields the plugin is not aware of.
- A workflow implementation must not react on *no-op* states nor terminal states.
  - The transition from a *no-op* state must be handled either by thin-edge as direct transition
    or overridden by the user with domain-specific checks and actions.
  - The terminal states, a.k.a `Successful` and `Failed`, are owned by the process which created the `Init` state (in practice, the mapper).
    Only this process should clear the retained message state for an operation instance by sending an empty payload on command's topic.

### Workflow Overriding

Thin-edge provides a mechanism to override, extend and combine workflows.

> This mechanism is provided by the `tedge-agent` which gather a set of user-defined workflows

Reviewer 2: Does  it mean that the `tedge-agent` is the owner of the full workflow?

Author: No. It means that `tedge-agent` manage the coordination and triggers the actions owned by tedge as running scripts and sending events.

> This mechanism is provided by the `tedge-agent` which gather a set of user-defined workflows

Author: This is the tricky point!

- First, files are not expected to be overridden. To override a workflow a new file is added for the same operation and target.
- In the [POC](https://github.com/didier-wenzek/tedge-mqtt-state-machine) the built-in workflows don't have to provide a file. The workflow manager and the built-in workflows are in this POC implemented as actors running in the *same* executable using the [`register_operation_plugin()`](https://github.com/didier-wenzek/tedge-mqtt-state-machine/blob/main/src/operations_sm/builder.rs#L46) method to exchange the implemented workflow.
- For this proposal, we can take the same approach. I.e. supporting each builtin operation with a `tedge_xxx_ext` actor spawned in the `tedge-agent` and not with an external `tedge_xxx_plugin` daemon.
- A `tedge_xxx_plugin` can be run along the `tedge-agent` as long as its workflow is described using a file given to the `tedge-agent`. However, there is a caveat: this plugin *must* ignore all the no-op states, otherwise the user would not be able to provide an alternative action, but this implies that the `tedge-xxx-plugin` will no more be able to run without the agent (the latter having to handle the no-op states or to delegate their execution to some user script). I still to find a good answer to that issue.

and combined them with the builtin workflows implemented by the agent itself and the thin-edge operation plugins.

Each workflow is defined using a TOML file that specifies:
- the operation target and name defined as a topic filter
  such as `te/device/main///cmd/software-update/+` or `te/device/main/service/+/cmd/restart/+`
- the list of states
- for each state:
  - the state name as defined by the operation API
  - the name of the state owner which is the process that is responsible for advancing a command at this state
    (the default being `tedge`, meaning that thin-edge is responsible for running the state checks and actions)
  - the set of states which can be an outcome for this state actions
  - possible extra instructions on how to process the command at this stage, e.g.
    - run a script
    - emit an event
    - raise an alarm
  

> ```toml title="file: workflow-example.toml"

Author: Added

target = "te/device/main///cmd/configuration-update/+"

[Init]
  next = ["Download", "Failed"]
  script = "/bin/schedule-configuration-update.sh"
    
[Download]

>   owner = tedge-configuration-plugin

Reviewer: ```suggestion
  owner = "tedge-configuration-plugin"
```

Author: fixed

>   owner = tedge-configuration-plugin

Author: Indeed in the [POC](https://github.com/didier-wenzek/tedge-mqtt-state-machine) nothing is done if the owner is not `tedge`.

However, if the MQTT identifier of this owner is given (i.e. `device/main/service/tedge-configuration-plugin`) a sophisticated  implementation could use this to check if thing has a chance to work.

=> This needs to be clarified

  next = ["Downloaded", "Failed"]

[Downloaded]
  script = "/bin/check-configuration-file.sh"
  next = ["Install", "Failed"]

[Install]

>   owner = "tedge-configuration-plugin"

Reviewer 3: Provided `tedge-configuration-plguin` is a daemon, this `owner` field is just indicating the daemon name for logging purpose or like that? I assume that the state machine doesn't really care about the process owner?

Author: Something still needs to be clarified.

Indeed, the `tedge-configuration-plugin` can be implemented as a:
- as a daemon - as of today
- as a script - as you suggested [here](https://github.com/thin-edge/thin-edge.io/pull/2071#discussion_r1268196345)
- as an actor running inside the agent - as implemented by the [POC](https://github.com/didier-wenzek/tedge-mqtt-state-machine)

These 3 alternatives have different pros/cons as well as implementation impacts. I will dive into these.

>   owner = tedge-configuration-plugin

Reviewer: ```suggestion
  owner = "tedge-configuration-plugin"
```

Author: fixed

  next = ["Installed", "Failed"]
        
[Installed]
  script = "/bin/check-configuration-update.sh"
  next = ["Successful", "Failed"]

[Successful]

>   event = "te/device/main///e/operations"

Reviewer: It might be worthwhile considering supporting a way to refer to the target topic that the workflow instance was triggered on. This would support writing more generic workflows (if the user chooses to do this).

For example if the workflow is configured to react to any `configuration-update` command, however they would like for the event to be created on the target associated to the command, then the `event` field in the `[Success]` state would need some way of referring to the target topic.

```toml
target = "te/+/+/+/+/cmd/configuration-update"
// ...
[Success]
  event = "{identifier}/e/operations"
```

  next = []
  
[Failed]
  alarm = "te/device/main///a/operations"
  next = []
```

> 

Author: I would prefer to be explicit and let the definition provide all the states.

Thin-edge combines all these workflows to determine what has to be done
when a state message is published for a command on one topic matching the global topic filter for commands,
i.e. `te/+/+/+/+/cmd/+/+`. Given an actual command state message, thin-edge 
- keeps only the workflows which target topic filter matches the topic of the message
- then keeps only the workflows that define a behavior for the state of the message payload
- and then uses priority rules to select a single workflow: the workflow ruling this command at that stage

> The priority rules give a higher priority to the workflow that are user-defined than to those pre-defined by thin-edge.

Reviewer: I like the concept of supporting priorities, to allow overriding workflows, and it is done in a very "linux friendly" format.

If several user-defined workflows are matching a command state,
then the alphabetic order of the workflow definition file names is used: 

> `001-configuration-update.toml` being of higher priority than `002-configuration-update.toml`.

Author: > For example, with the `config-update` operation, there'd be different kinds of custom scripts to perform updates of different types of configs.

The proposal is to have a *single* set of actions for a (command, status) pair. If you want to behave differently depending on some property of the state payload (here the config type) then this has to be handle *inside* the script. If you want to handle this at the workflow level, you can introduce new state (e.g. `update_mosquitto_config` and `update_other_config`) but you need then to run a script on the state `update_config` that move to one or the other state depending on the config type.

> 

Reviewer 2: Can there be a parallel state in a workflow?

Author: If by parallel state you mean that two states can be observed at the same time, then no.

---
title: Operation Workflows
tags: [Reference, MQTT, Operations]
sidebar_position: 6
---

# Operation Workflows

Thin-edge provides the tools to define, extend and combine *operation workflows*
that rule the sequence of steps applied when a maintenance *operation* is triggered by an operator or some software component,
whether it is a *command* to restart the device, to update a configuration file or to install a new software.

An operation workflow specifies the timeline of the actions to perform
as well as any prerequisite checks, outcome validations and possible rollbacks.
However, a workflow doesn't define how to perform these actions.
These are delegated to software components participating in the operation progress.
More precisely, an operation workflow defines:
- the *observable states* of an ongoing operation instance
  from initialization up to a final success or failure
- the *participants* and their interactions, passing the baton to the software component

>   which responsibility is to advance the operation in a given state

Author: fixed

Here, the `tedge-agent`, running on the main device, is requested for a health status with the following command:

```sh te2mqtt
tedge mqtt pub 'te/device/main/service/tedge-agent/cmd/heath-check/req' '{}' 
```

Software update is an example of an operation that requires coordination.
On reception of a software update request from the cloud operator,
the `tedge-mapper` creates a fresh new topic for this command,
say `te/device/main///cmd/software-update/req-123` for the 123<sup>rd</sup> request.
On this topic, a first retained messages is published to describe the operator expectations for the software updates.

```sh te2mqtt
tedge mqtt pub -r 'te/device/main///cmd/software-update/req-123' '{
    "status": "init",
    "modules": [
        {
            "type": "apt",
            "name": "collectd",
            "version": "5.7",
            "action": "install"
        }
    ]
}' 
```

Then, the `tedge-agent` and possibly other software components take charge of the command, making it advance to some final state,
publishing all the successive states as retained messages on the command topic.

> Eventually, the `tedge-mapper` will have to clean the command topic with an empty retained message: 

Author: Publishing an empty retained message is the MQTT way to clear the state data persisted along a topic.

  - Such a message states where the command progression is and gives all the required information to proceed.
  - e.g. `{ "status": "Requested", "target": "mosquitto", "url": "https://..." }`
- The state messages are published as retained.
  - They capture the latest state of the operation request.
  - Till some change occurs, this latest state is dispatched to any participant on reconnect.
- Several participants act in concert to move forward the command execution.
  - The participants observe the progress of all the operations they are interested in.
  - They watch for the specific states they are responsible in moving forward.
  - When a step is performed, successfully or not, the new state is published accordingly by the performer.

### Example

Here is an example where three software components participate in a `configuration-update` command.
- The `tedge-mapper` creates the initial state of the command
  providing the required information to install a new version for a configuration file;
  and then waits for the final outcome (in black).
- The `tedge-config-plugin` handles the main steps (in red): downloading the file and installing it where expected.

> - An extra component, provided by the user, handles domain-specific checks (in blue)

Author: > does the 'tedge-config-plugin' wait for the "Downloaded" state to be published on the topic by 'extra-component'so that it can do the "install" step?

Each plugin will only react on specific states, ignoring all others.

> This will be aggrated in the reverse situation- which means i cannot add an additional step in workflow without needing to change the exsisting logic

This is a strong requirement for these workflows: "a user should be able to extend an operation workflow, *without* changing the logic provided out-of-the-box by thin-edge".

In its current version, this document proposes to introduce extra steps as `Downloaded` to give the user an opportunity to do something extra between the `Downloading` and `Installing` steps.

That said many things still need to be clarified. The main issues being to ensure there one and only one plugin responsible of moving forward a command in its current state.

- user documentation of the required input and the expected outcome of an operation request
- the set of observable states for a command and the possible state sequences
- for each state:
  - the well-known name such as `Download` or `Downloaded`
  - the schema of the state payload and the required parameters to process a command at this stage 
  - developer documentation on the role of each parameter and the expected checks and actions
- the schema for the capability message sent when the operation is enabled on some thin-edge entity or component
  - developer documentation on the role of each field of the capability message

A workflow implementation is free to define the states a command can go through
as well as the message payload attached to each state.

However, there are some rules and best practices.

- Three states are mandatory: `Init`, `Successful` and `Failed`.
- `Init` must be the unique initial state of the workflow.
  - This state must give all the information required to execute the command.
  - Having no other initial states is important so any command requester (as the mapper)
    can trigger commands in a systematic manner *even* if the workflow is updated.

> - `Successful` and `Failed` must be the unique terminal states of the workflow.

Author: Indeed something needs to be clarified.

By `failed` terminal step, I mean a state for the final conclusion returned to the requester : its command failed. If some rollback actions have been performed, these must have been done *before*. In practice, this means that the workflow designer will have to introduce specific states to handle rollbacks. Some extra states along these lines: `executing` -> `rollback-required` -> `rollbacking` -> `failed`.

It's important to have well-known fixed terminal states, otherwise the mapper will have no way to know when the execution is actually finished.
